DataLayer/Wearable/src/main/java/com/example/android/wearable/datalayer/NodesViewModel.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.datalayer
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.android.wearable.datalayer.MainActivity.Companion.CAMERA_CAPABILITY
+import com.example.android.wearable.datalayer.MainActivity.Companion.MOBILE_CAPABILITY
+import com.example.android.wearable.datalayer.MainActivity.Companion.WEAR_CAPABILITY
+import com.google.android.gms.wearable.CapabilityClient
+import com.google.android.gms.wearable.Node
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.shareIn
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.tasks.await
+
+class NodesViewModel(
+    private val capabilityClient: CapabilityClient
+) : ViewModel() {
+
+    private val nodes: SharedFlow<Set<Node>> = flow {
+        emit(
+            getCapabilitiesForReachableNodes()
+                .filterValues { MOBILE_CAPABILITY in it || WEAR_CAPABILITY in it }.keys
+        )
+    }.shareIn(
+        viewModelScope,
+        started = SharingStarted.Eagerly,
+        replay = 1
+    )
+
+    private val cameraNodes: SharedFlow<Set<Node>> = flow {
+        emit(
+            getCapabilitiesForReachableNodes()
+                .filterValues { MOBILE_CAPABILITY in it && CAMERA_CAPABILITY in it }.keys
+        )
+    }.shareIn(
+        viewModelScope,
+        started = SharingStarted.Eagerly,
+        replay = 1
+    )
+
+    /**
+     * Collects the capabilities for all nodes that are reachable using the [CapabilityClient].
+     *
+     * [CapabilityClient.getAllCapabilities] returns this information as a [Map] from capabilities
+     * to nodes, while this function inverts the map so we have a map of [Node]s to capabilities.
+     *
+     * This form is easier to work with when trying to operate upon all [Node]s.
+     */
+    private suspend fun getCapabilitiesForReachableNodes(): Map<Node, Set<String>> =
+        capabilityClient.getAllCapabilities(CapabilityClient.FILTER_REACHABLE)
```

Comment:
This call can fail, so we should probably catch the ApiExceptions and log?

Reply:
Under which circumstances can fail?

Reply:
It's not likely, more about correct code.

https://developers.google.com/android/reference/com/google/android/gms/wearable/WearableStatusCodes

I think any call can technically fail with TIMEOUT, REMOTE_EXCEPTION, INTERNAL_ERROR.

#tioli

Reply:
We should consider these a remote call to another process on the device.